openfisca_core/holders.py

```diff
@@ -112,86 +110,72 @@ def clone(self):
         """
         if period is None:
             period = self.simulation.period
         column = self.column
-        accept_other_period = parameters.get('accept_other_period', False)
 
-        # First look for dated_holders covering the whole period (without hole).
+        # Check that the requested period matches period_behavior
+        if not self.column.period_behavior == PERMANENT:
+            if ((self.column.period_behavior == MONTH and period.unit != periods.MONTH) or
+                    (self.column.period_behavior == YEAR and period.unit != periods.YEAR)):
+                raise ValueError('Computation requested with wrong period unit for variable {} ({} instead of {})'.format(
+                    self.column.name,
+                    period.unit, self.column.period_behavior))
+        if period.size != 1:
+            raise ValueError('Computation requested for complex period {} for variable {}'.format(
+                period, self.column.name))
+
+        # First look for a value already cached
         dated_holder = self.get_from_cache(period, parameters.get('extra_params'))
         if dated_holder.array is not None:
             return dated_holder
         assert self._array is None  # self._array should always be None when dated_holder.array is None.
 
+        # Request a computation
         column_start_instant = periods.instant(column.start)
         column_stop_instant = periods.instant(column.end)
         if (column_start_instant is None or column_start_instant <= period.start) \
                 and (column_stop_instant is None or period.start <= column_stop_instant):
             formula_dated_holder = self.formula.compute(period = period, **parameters)
             assert formula_dated_holder is not None
             if not column.is_permanent:
-                assert accept_other_period or formula_dated_holder.period == period, \
+                assert formula_dated_holder.period == period, \
                     "Requested period {} differs from {} returned by variable {}".format(period,
                         formula_dated_holder.period, column.name)
             return formula_dated_holder
         array = self.default_array()
         return self.put_in_cache(array, period)
 
     def compute_add(self, period = None, **parameters):
-        dated_holder = self.get_from_cache(period, parameters.get('extra_params'))
-        if dated_holder.array is not None:
-            return dated_holder
+        # Check that the requested period matches period_behavior
+        if self.column.period_behavior == YEAR and period.unit == periods.MONTH:
+            raise ValueError('Computation on period {} impossible on yearly variable {}'.format(
+                period, self.column.name
+                ))
+
+        if self.column.period_behavior == MONTH:
+            variable_period_unit = periods.MONTH
+        elif self.column.period_behavior == YEAR:
+            variable_period_unit = periods.YEAR
```

Comment:
Even though @benjello's point about a lower point is totally valid from an analytical standpoint, I would still recommend `ETERNITY` for consistency with other enum names.

If the other types are `MONTH` and `YEAR`, for example, you want the last one to also be a noun, not an adjective (`PERMANENT`) or an adverb (`ALWAYS`).

This might sound like arguing over nothing at first, but I can tell you it will otherwise always have that little weird feeling when you type it, even though you might not realise why, and your brain will always be slightly longer to retrieve the word (until you're so used to it it doesn't carry any meaning anymore —but API design should target newcomers too :wink:).